demo-circuit/src/sc2sc/mod.rs

```diff
@@ -0,0 +1,316 @@
+    {
+        GingerMHTBinaryGadget::alloc(cs.ns(|| "alloc messages tree path"), || {
+            Ok(self.msg_path.clone())
+        })?
+        .enforce_root_from_leaf(
+            cs.ns(|| "reconstruct_merkle_root_hash(msg_hash, msg_path)"),
+            msg_hash_g,
+        )?
+        .enforce_equal(
+            cs.ns(|| {
+                "Verify merkle root of (msg_hash, msg_path) == curr_cert.SC2SC_message_tree_root"
+            }),
+            msg_root_g,
+        )?;
+        Ok(())
+    }
+}
+
+impl ConstraintSynthesizer<FieldElement> for Sc2Sc {
+    fn generate_constraints<CS: r1cs_core::ConstraintSystemAbstract<FieldElement>>(
+        self,
+        cs: &mut CS,
+    ) -> Result<(), r1cs_core::SynthesisError> {
+        // Expose public inputs
+        let next_sc_tx_commitments_root_g =
+            FieldElementGadget::alloc_input(cs.ns(|| "Alloc next sc tx commitment root"), || {
+                Ok(self.next_sc_tx_commitments_root)
+            })?;
+        let curr_sc_tx_commitments_root_g = FieldElementGadget::alloc_input(
+            cs.ns(|| "Alloc current sc tx commitment root"),
+            || Ok(self.curr_sc_tx_commitments_root),
+        )?;
+        let msg_hash_g =
+            FieldElementGadget::alloc_input(cs.ns(|| "Alloc msg_hash"), || Ok(self.msg_hash))?;
+
+        // Alloc Withdraw certificate gadgets
+        let next_cert_g =
+            WithdrawalCertificateDataGadget::alloc(cs.ns(|| "alloc next wcert data"), || {
+                Ok(self.next_cert.clone())
+            })?;
+        let curr_cert_g =
+            WithdrawalCertificateDataGadget::alloc(cs.ns(|| "alloc current wcert data"), || {
+                Ok(self.curr_cert.clone())
+            })?;
+        // Enforce certificates hashes
+        let curr_cert_hash_g =
+            curr_cert_g.enforce_hash(cs.ns(|| "enforce current certificate hash"), None)?;
+        let next_cert_hash_g =
+            next_cert_g.enforce_hash(cs.ns(|| "enforce next certificate hash"), None)?;
+
+        // Enforce next certificate top quality hash == current certificate hash
+        next_cert_g.custom_fields_g[MAX_QUALITY_CERT_HASH_CUSTOM_FIELDS_POS].enforce_equal(
+            cs.ns(|| "require(next_cert.previous_top_quality_hash == H(curr_cert_hash)"),
+            &curr_cert_hash_g,
+        )?;
+
+        // Enforce current certificate epoch id + 1 == next certificate epoch id
+        self.enforce_contiguos_epochs(cs, &curr_cert_g, &next_cert_g)?;
+
+        // Enforce merkle paths for current certificate, next certificate and message hash
+        self.curr_cert_path.enforce_sc_tx_commitment_root(cs.ns(|| "Current epoch sc_tx_commitment_root recostruction") , 
+            curr_cert_hash_g, curr_cert_g.ledger_id_g)?
+            .enforce_equal(
+                cs.ns(|| "Verify merkle root of (curr_sc_tx_commitment, curr_sc_tx_commitments_path) == curr_sc_tx_commitments_root"),
+                &curr_sc_tx_commitments_root_g,
+            )?;
```

Comment:
It was an explicit choice. Before I wrote it as you propose now and after I changed it because it seams to me a little bit more readable.
... but maybe if we use `check_membership` is better to put it inside ... I don't like to pass too much argument to a function.

Reply:
See above comment... I called it `check_membership()` because it have the responsibility  f the merkle path one.